### PR TITLE
feat: add custom objective function to lightgbm learners

### PR DESCRIPTION
**Pull request by imatiach-msft**
add support for custom objective function in lightgbm learners

Users in the scala API will now be able to supply a custom loss function for training, which will need to output the gradient and hessian matrix.

As part of this PR, the LightGBMBooster and LightGBMDataset classes have been improved and moved under the booster and dataset folders, and some of the files related to parameters have been moved to a separate params folder in order to organize the code better since there are so many files in lightgbm directory.

TODO: add pyspark support, in a separate PR

### CODE REVIEW

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala [hunk 1]
```diff
@@ -167,35 +200,129 @@ protected class BoosterHandler(model: String) {
...
+      newParameters), "Booster Reset learning_rate Param")
+  }
+
+  def innerPredict(dataIndex: Int, classification: Boolean): Array[Array[Double]] = {
+    val kind = boosterHandler.normalScoreConstant
+    lightgbmlib.LGBM_BoosterGetPredict(boosterHandler.boosterPtr, dataIndex,
+      boosterHandler.customObjectiveScoredDataLengthLongPtr.get().ptr,
+      boosterHandler.customObjectiveScoredDataOutPtr.get().ptr)
+    // TODO: fixme
+    Array(predScoreToArray(classification, boosterHandler.customObjectiveScoredDataOutPtr.get().ptr, kind))
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:323]: TODO: fixme

File: src/test/scala/com/microsoft/ml/spark/lightgbm/split1/VerifyLightGBMClassifier.scala [hunk 2]
```diff
@@ -86,6 +87,7 @@ trait LightGBMTestUtils extends TestBase {
   }
 
   def assertProbabilities(tdf: DataFrame, model: LightGBMClassifier): Unit = {
+    println(model.getModelString)
```
Comment (imatiach-msft) [inline src/test/scala/com/microsoft/ml/spark/lightgbm/split1/VerifyLightGBMClassifier.scala:90]: remove

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala (hunk 1 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala]: done

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala [hunk 3]
```diff
@@ -1,12 +1,15 @@
 // Copyright (C) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in project root for information.
 
-package com.microsoft.ml.spark.lightgbm
+package com.microsoft.ml.spark.lightgbm.booster
```
Comment (mhamilton723) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:4]: why the namespace change?

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala [hunk 4]
```diff
@@ -167,35 +182,155 @@ protected class BoosterHandler(model: String) {
...
+    if (trainDataset.isEmpty) {
+      new BoosterHandler(model)
+    } else {
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterCreate(trainDataset.map(_.datasetPtr).get,
+        parameters.get, boosterOutPtr), "Booster")
+      new BoosterHandler(lightgbmlib.voidpp_value(boosterOutPtr))
+    }
   }
 
+  val model = modelStr.getOrElse("")
```
Comment (mhamilton723) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:230]: what is the behavior if this is ""? Seems this should be kept as option especially if code elsewhere checks for ""

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala [hunk 5]
```diff
@@ -167,35 +182,155 @@ protected class BoosterHandler(model: String) {
...
 /** Represents a LightGBM Booster learner
-  * @param model The string serialized representation of the learner
+  * @param trainDataset The training dataset
+  * @param parameters The booster initialization parameters
+  * @param modelStr Optional parameter with the string serialized representation of the learner
   */
 @SerialVersionUID(777L)
-class LightGBMBooster(val model: String) extends Serializable {
-  /** Transient variable containing local machine's pointer to native booster
+class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val parameters: Option[String] = None,
```
Comment (mhamilton723) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:197]: will this break serialization?

**Review by mhamilton723 — DISMISSED** [review]

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala (hunk 3 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:4]: moving params, booster and dataset into their own folders, because there are way too many files in lightgbm folder

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala (hunk 4 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:230]: model string should never be empty

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala (hunk 5 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala:197]: it's still Serializable, see line below

File: src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala (hunk 4 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala]: removed the model

File: src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+trait FObjTrait extends Serializable {
+  /**
+    * User defined objective function, returns gradient and second order gradient
+    *
+    * @param predictions untransformed margin predicts
+    * @param trainingData training data
+    * @return List with two float array, correspond to grad and hess
+    */
+  def getGradient(predictions: Array[Array[Double]], trainingData: LightGBMDataset): List[Array[Float]]
```
Comment (mhamilton723) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala:16]: Might want to make this return two objects instead of a list. Also in general the grad and hessian wont be the same shape...

File: src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.params
+
+import com.microsoft.ml.spark.lightgbm.dataset.LightGBMDataset
+
+trait FObjTrait extends Serializable {
```
Comment (mhamilton723) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala:8]: Perhaps we should rename this to CustomObjective

**Review by mhamilton723 — CHANGES_REQUESTED** [review]

File: src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala (hunk 7 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala:8]: I think that's what it is called in XGBoost actually, but in lightgbm the parameter is called fobj, so I wanted to keep in line with the terminology in lightgbm

File: src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala (hunk 6 above)
Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala:16]: interesting idea, in xgboost I see they return one list:
https://github.com/dmlc/xgboost/blob/4224c08cacceba3f83f90e387c07aa6205a83bfa/jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/ObjectiveTrait.scala#L32
I can return two objects but for those familiar with xgboost it might be harder to get used to a slightly different API

Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala]: done!  changes it to return a tuple

File: src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala (hunk 7 above)
Comment (mhamilton723) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/params/FObjTrait.scala:8]: SG

**Review by mhamilton723 — DISMISSED** [review]
